client/src/margo_client.c

```diff
@@ -286,6 +296,7 @@ int invoke_client_metaset_rpc(unifyfs_file_attr_t* f_meta)
 
     /* fill in input struct */
     unifyfs_metaset_in_t in;
+    in.create       = (int32_t) create;
```

Comment:
Question: what if you didn't have a `in.create` field, and simply created the file if it didn't exist?  Or are there times you would want to either call metaset on a non-existant file, or create=1 on an existing file?

Reply:
Currently, invoke_client_metaset_rpc is serving two roles as both our "file create" and "update metadata" rpc depending on whether the user is specifying create=1 or not.

With create=1, the server updates all metadata in the global key/value store, including the file size and is_laminated flags.  It creates the entry if it doesn't exist, and it silently overwrite any existing entry.

With create=0, the server will preserve the size and is_laminated fields for any existing entry in the global/key value store, but it updates all other fields.  If there is no existing entry already, it uses the size/is_laminated values provided by the caller.

For a true "create" operation, it would probably be desirable to have the server return an error if the item already exists.  However, we have a problem in that our key/value store does not yet support atomic read/modify/write operations, which we'd need to safely enable this.

Reply:
After double-checking the code, with create=0, it returns an error from the set metadata rpc if it does not find an existing entry, which is good.

Reply:
Does it also error if you do create=1 when the file already exists?

Reply:
No, it behaves like an O_CREAT rather than an "O_CREAT | O_EXCL".  There is some other code within the server that depends on that at the moment.

I can imagine we'll want to be able to have both options but I'll save that extension for a later PR, especially since we can't do atomic creation operations with the current key/value store.

Reply:
Ok.  Is there any way we can make it clearer/simplify what metaset_rpc does?  Proposal:
1. Remove `unifyfs_metaset_in_t.is_laminated`, to explicitly disallow setting `is_laminated` from metaset_rpc.  That way, there's only one way to laminate (via laminate_rpc) and we can't shoot ourselves in the foot.

2. On the server, always update the file size on a metaset_rpc, *except* if the file is laminated.  Return an error in that case.

3. Add a more detailed comment describing the function:
```C
/* 
 * Set the metadata values for a file (after optionally creating it).  The
 * gfid for the file is in f_meta->gfid.
 * 
 * create:	If set to 1, attempt to create the file first.  If the file
 *              already exists, then update its metadata with the values in
 *              f_meta.  If set to 0, and the file does not exist, then
 *              the server will return an error.
 *
 * f_meta:	The metadata values to update.
 */
 ```

Reply:
Thanks for the good function description.  I added that to margo_client.c file.  The header is comment-free at the moment, so putting it in the header makes me feel like I should go comment all of the functions there.

Before submitting this PR, I had considered removing the size and is_laminated fields from the metaset rpc input parameter list.  However, I thought there could be some case in the future where being able to set those fields from the client would be helpful.  I decided to leave them in for that reason.  I feel like the real solution would be to bite the bullet and define an actual mknod_rpc that would be the pair of the unlink_rpc.  Then we could make the metaset_rpc only update existing items and not create new ones.

Also for the metaset rpc, as @MichaelBrim was suggesting, we'll eventually want to be able to selectively update particular fields (e.g., change the gid but nothing else).  For that, we'd want to include some sort of bit flags in the input parameters or something...  That's more than I wanted to get into here.  For this PR, I just want metaset to stop changing the size field.